Whole-game.Rmd

```diff
@@ -109,300 +125,375 @@ create <- !is.na(where)
 
-Click on History (the clock icon) and, if you consented, you will see an initial commit made via `use_git()`:
+Click on History (the clock icon in the Git pane) and, if you consented, you will see an initial commit made via `use_git()`:
 
-:::downlit
 ```{r inspect-first-commit, echo = FALSE, eval = create}
 git_log(max = 1) %>% 
   select(commit, author, message)
 ```
-:::
 
 :::rstudio-tip
-RStudio can initialize a Git repository, in any Project, even if it's not an R package, as long you've set up RStudio + Git integration. Do *Tools > Version Control > Project Setup*. Then choose *Version control system: Git* and *initialize a new git repository for this project*.
+RStudio can initialize a Git repository, in any Project, even if it's not an R package, as long you've set up RStudio + Git integration.
+Do *Tools > Version Control > Project Setup*.
+Then choose *Version control system: Git* and *initialize a new git repository for this project*.
 :::
 
 ## Write the first function
 
-It is not too hard to find a puzzling operation involving factors. Let's see what happens when we catenate two factors.
+A fairly common task when dealing with strings is the need to split a single string into many parts.
+The `strsplit()` function in base R does exactly this.
 
 ```{r}
-(a <- factor(c("character", "hits", "your", "eyeballs")))
-(b <- factor(c("but", "integer", "where it", "counts")))
-c(a, b)
+(x <- "alfa,bravo,charlie,delta")
+strsplit(x, split = ",")
 ```
 
-NOTE: As of R 4.1.0 (released 2021-05-18), "using `c()` to combine a factor with other factors now gives a factor", instead of an integer vector.
-Hallelujah!
-Here's what `c(a, b)` looked like in R until May 2021, i.e. in R < 4.1.0:
+Take a close look at the return value.
 
-```{r eval = FALSE}
-c(a, b)
-#> [1] 1 3 4 2 1 3 4 2
+```{r}
+str(strsplit(x, split = ","))
 ```
 
-NOTE continued: What puzzling thing about R should be the basis of a new toy package example? Who knows, featuring it here might help get it fixed in base R itself! You can record your ideas in this GitHub issue: <https://github.com/hadley/r-pkgs/issues/709>.
+The shape of this return value often surprises people or, at least, inconveniences them.
+The input is a character vector of length one and the output is a list of length one.
+This makes total sense in light of R's fundamental tendency towards vectorization.
+But sometimes it's still a bit of a bummer.
+Often you know that your input is morally a scalar, i.e. it's just a single string, and really want the output to be the character vector its parts.
 
-Huh? Many people do not expect the result of catenating two factors to be an integer vector consisting of the numbers 1, 2, 3, and 4. What if we coerce each factor to character, catenate, then re-convert to factor?
+This leads R users to employ various methods of "unlist"-ing the result:
 
 ```{r}
-factor(c(as.character(a), as.character(b)))
+unlist(strsplit(x, split = ","))
+
+strsplit(x, split = ",")[[1]]
```

Comment:
I am half-seriously considering doing some sort of search to grab lines containing `strsplit()` in R code on GitHub and do a rough analysis of what proportion are one of these forms. I feel like I've even done this in the past and the proportion is quite high! Would be good for a footnote at least.

Reply:
Informally, if I do a GitHub search for "strsplit path:R user:cran", 9 of the 10 hits on the first page use either `unlist()` or `[[1]]`. The remaining one uses `[[1]]` when the result is actually used in the next line. So 10 for 10 of these examples actually want the un-listed result 🤣